decisions/01-new-animation-api.md

```diff
@@ -0,0 +1,208 @@
+
+We propose a new animation API where design functions default to a frame-based approach. If a design function is animated, it will automatically get called continously in a pace determined by the `frameRate` setting, until it calls the `done` function to stop the animation. This means that design functions are expected to be pure functions and not rely on global state to render the given frame. This in turn makes it possible for the Mechanic UI to provide a timeline that will update the design function with the given `frameCount` and result in a preview for that specific frame. Note that this new animation API does not impose any rules about the duration of the animation. This is always controlled by calling `done` at some point during the animation lifecycle.
+
+## Implementation Details
+
+The new animation API comes with a few changes to the settings:
+
+- The `animated` setting is renamed to `mode` (default: `static`) and the options are `static` (for static images), `animation` (for the new animation API) and `customAnimation` (for a user to implement their own animation code using the `frame` and `done` callbacks).
+- `frameRate` (default: `60`) is a number that can be used to change the number of frames per second that the design function is called in `animation` mode.
+
+Since the new animation API encourages pure functions, the engines now ship with a `memo` function that can be used to cache things across function calls. Engine-specific helpers such as a `useCanvas` function will also be rewritten to create the canvas element on the first call and return it on following calls.
+
+The argument syntax is also changing slightly by placing the `frame` and `done` functions as root properties of the design function argument object. The `frame` function will only get passed to the design function in `customAnimation` mode.
+
+Here's a look at what this new animation API will look like for each engine.
+
+### `engine-canvas`
+
+```js
+export const handler = async ({ inputs, frameCount, done, useCanvas }) => {
+  const canvas = useCanvas(inputs.width, inputs.height);
+  if (frameCount >= 100) {
+    done(canvas);
+  }
+  // drawing code
+  return canvas;
+};
+
+export const settings = {
+  engine: require('@mechanic-design/engine-canvas'),
+  mode: 'animation'
+};
+```
+
+```js
+export const handler = async ({ inputs, frame, done, useCanvas }) => {
+  const canvas = useCanvas(inputs.width, inputs.height);
+  let frameCount = 0;
+  const myDrawLoop = () => {
+    frameCount++;
+    // drawing code
+    if (frameCount >= 100) {
+      done(canvas);
+    } else {
+      frame(canvas);
+      window.requestAnimationFrame(myDrawLoop);
+    }
+  };
+  myDrawLoop();
+};
+
+export const settings = {
+  engine: require('@mechanic-design/engine-canvas'),
+  mode: 'customAnimation'
+};
+```
+
+### `engine-svg`
+
+```js
+export const handler = async ({ inputs, frameCount, done }) => {
+  // drawing code
+  if (frameCount >= 100) {
+    done(svgString);
```

Comment:
Just thinking about this. This forces users to keep `svgString` in a variable so it can either be returned or be send to `done`. While this isn't necessarily a problem as you have to the same using `frame` and `done` currently I think it might feel weird as `return` encourages to just build the return value there (at least to me).

So maybe the API could be simplified, that in `animation` mode `done` just flips a boolean in the engine that tells the engine to stop rendering after the next return value is received.

Reply:
I was actually just thinking about this. I think there's some weirdness around whether the user needs to pass something to the callbacks or not. So I like the idea of `animation` removing the need to do that and `customAnimation` requiring you to do that. We just need to keep in mind the difference that in `animation` mode, the `done` function will be called before the return, thus the last frame will not be recorded as a part of the animation, whereas it will be when you pass the string to the `done` function in `customAnimation`.

Reply:
That, or we change it so the `done` function never receives anything and always ends the animation immediately. That way, the signature for the `frame` and the `done` function will stay consistent, and the different modes are simply about whether you use the `frame` function or not. I think I like that slightly better.

Reply:
Agreed